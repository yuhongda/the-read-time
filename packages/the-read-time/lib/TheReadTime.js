"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.TheReadTime = void 0;
var _react = _interopRequireDefault(require("react"));
var _wordCount = _interopRequireDefault(require("./wordCount"));
var _extractTextFromReactNode = _interopRequireDefault(require("./extractTextFromReactNode"));
var _misc = require("./misc");
var _convertTimeToEmoji = _interopRequireDefault(require("./convertTimeToEmoji"));
var _convertNumberToTime = _interopRequireDefault(require("./convertNumberToTime"));
var _excluded = ["steps", "speed", "time", "displayRender"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var TheReadTime = function TheReadTime(props) {
  var _props$steps = props.steps,
    steps = _props$steps === void 0 ? _misc.DEFAULT_STEPS : _props$steps,
    _props$speed = props.speed,
    speed = _props$speed === void 0 ? _misc.DEFAULT_READ_SPEED : _props$speed,
    time = props.time,
    displayRender = props.displayRender,
    rest = _objectWithoutProperties(props, _excluded);
  var totalTime;
  if (time) {
    totalTime = time;
  } else {
    var text = (0, _extractTextFromReactNode["default"])(props.children);
    var count = (0, _wordCount["default"])(text);
    totalTime = count / speed;
  }
  var totalTimeString = (0, _convertNumberToTime["default"])(totalTime);
  var processSteps = steps.sort(function (a, b) {
    return b.minutes - a.minutes;
  });
  var emoji = (0, _convertTimeToEmoji["default"])(processSteps, totalTime);
  return /*#__PURE__*/_react["default"].createElement("div", rest, displayRender ? displayRender(emoji, totalTime) : /*#__PURE__*/_react["default"].createElement("div", null, emoji.flat().join(''), totalTimeString), props.children);
};
exports["default"] = exports.TheReadTime = TheReadTime;