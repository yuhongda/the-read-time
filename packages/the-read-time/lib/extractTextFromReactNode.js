"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
var extractTextFromReactNode = function extractTextFromReactNode(node) {
  if (node === null || node === undefined) {
    return '';
  }
  if (typeof node === 'string') {
    return node;
  }
  if (typeof node === 'number') {
    return "".concat(node);
  }
  if (typeof node === 'boolean') {
    return '';
  }
  if (Array.isArray(node)) {
    return node.map(function (n) {
      return extractTextFromReactNode(n);
    }).join('');
  }
  if ( /*#__PURE__*/_react["default"].isValidElement(node) && node.hasOwnProperty('props')) {
    if (node.props.hasOwnProperty('children')) {
      return extractTextFromReactNode(node.props.children);
    }
    if (node.props.hasOwnProperty('dangerouslySetInnerHTML') && node.props.dangerouslySetInnerHTML.hasOwnProperty('__html')) {
      return extractTextFromReactNode(node.props.dangerouslySetInnerHTML.__html);
    }
  }
  return '';
};
var _default = extractTextFromReactNode;
exports["default"] = _default;