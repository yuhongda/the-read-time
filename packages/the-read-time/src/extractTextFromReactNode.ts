import type { ReactNode } from 'react'
import React from 'react'

const extractTextFromReactNode = (node: ReactNode): string => {
  if (node === null || node === undefined) {
    return ''
  }

  if (typeof node === 'string') {
    return node
  }

  if (typeof node === 'number') {
    return `${node}`
  }

  if (typeof node === 'boolean') {
    return ''
  }

  if (Array.isArray(node)) {
    return node.map((n) => extractTextFromReactNode(n)).join('')
  }

  if (React.isValidElement(node) && node.hasOwnProperty('props')) {
    if (node.props.hasOwnProperty('children')) {
      return extractTextFromReactNode(node.props.children)
    }
    if (
      node.props.hasOwnProperty('dangerouslySetInnerHTML') &&
      node.props.dangerouslySetInnerHTML.hasOwnProperty('__html')
    ) {
      return extractTextFromReactNode(node.props.dangerouslySetInnerHTML.__html)
    }
  }

  return ''
}

export default extractTextFromReactNode
