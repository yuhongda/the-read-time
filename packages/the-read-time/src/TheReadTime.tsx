import React from 'react'
import type { TheReadTimeProps } from './types'
import wordCount from './wordCount'
import extractTextFromReactNode from './extractTextFromReactNode'
import { DEFAULT_READ_SPEED, DEFAULT_STEPS } from './misc'
import convertTimeToEmoji from './convertTimeToEmoji'
import convertNumberToTime from './convertNumberToTime'

export const TheReadTime: React.FC<TheReadTimeProps> = props => {
  const { steps = DEFAULT_STEPS, speed = DEFAULT_READ_SPEED, time, displayRender, ...rest } = props

  let totalTime
  if (time) {
    totalTime = time
  } else {
    const text = extractTextFromReactNode(props.children)
    const count = wordCount(text)
    totalTime = count / speed
  }
  const totalTimeString = convertNumberToTime(totalTime)
  const processSteps = steps.sort((a, b) => b.minutes - a.minutes)
  const emoji = convertTimeToEmoji(processSteps, totalTime)

  return (
    <div {...rest}>
      {displayRender ? (
        displayRender(emoji, totalTime)
      ) : (
        <div>
          {emoji.flat().join('')}
          {totalTimeString}
        </div>
      )}
      {props.children}
    </div>
  )
}
export default TheReadTime
