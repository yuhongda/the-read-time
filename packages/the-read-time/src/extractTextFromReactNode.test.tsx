import { describe, it, expect } from 'vitest'
import { render, screen, cleanup } from '@testing-library/react'
import extractTextFromReactNode from './extractTextFromReactNode'
import React from 'react'
import TheReadTime from './TheReadTime'

describe('extractTextFromReactNode()', () => {

  it('extractTextFromReactNode works fine', async () => {
    const text = extractTextFromReactNode(
      <>
        <p>
          Many resources indicate that the average reading speed of most adults is around 200 to 250
          words per minute, which translates to two minutes per page. College students, probably
          because they must practice reading, move that pace up a notch to about 300 words per
          minute. Admittedly, not everyone can read fast and hit the average reading speed
          regardless of one’s age. Many factors play a role in how fast a person can read, such as
          practice, reading material, and their health. However, this does not mean that they can’t
          improve their reading pace. Tag along as we discuss average reading speed, factors
          affecting one’s reading speed and comprehension, and ways to improve reading speed.
        </p>
        The Article <strong>blablabla...</strong> ha-ha!! 哈哈哈
      </>
    )
    expect(text).toMatchSnapshot()
  })
})
