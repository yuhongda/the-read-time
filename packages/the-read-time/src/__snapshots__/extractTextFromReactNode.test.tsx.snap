// Vitest Snapshot v1

exports[`extractTextFromReactNode() > 100 minutes => 🍱🍱🍱🍱🍱🍱🍔🍔 1`] = `"Many resources indicate that the average reading speed of most adults is around 200 to 250 words per minute, which translates to two minutes per page. College students, probably because they must practice reading, move that pace up a notch to about 300 words per minute. Admittedly, not everyone can read fast and hit the average reading speed regardless of one’s age. Many factors play a role in how fast a person can read, such as practice, reading material, and their health. However, this does not mean that they can’t improve their reading pace. Tag along as we discuss average reading speed, factors affecting one’s reading speed and comprehension, and ways to improve reading speed.The Article blablabla... ha-ha!! 哈哈哈"`;

exports[`extractTextFromReactNode() > extractTextFromReactNode works fine 1`] = `"Many resources indicate that the average reading speed of most adults is around 200 to 250 words per minute, which translates to two minutes per page. College students, probably because they must practice reading, move that pace up a notch to about 300 words per minute. Admittedly, not everyone can read fast and hit the average reading speed regardless of one’s age. Many factors play a role in how fast a person can read, such as practice, reading material, and their health. However, this does not mean that they can’t improve their reading pace. Tag along as we discuss average reading speed, factors affecting one’s reading speed and comprehension, and ways to improve reading speed.The Article blablabla... ha-ha!! 哈哈哈"`;
